src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,497 @@
+symbol `jacobi_sym a b`. (Unfortunately, there is no subscript "J" in unicode.)
+-/
+
+open zmod nat
+
+/-- The Jacobi symbol of `a` and `b` -/
+-- Since we need the fact that the factors are prime, we use `list.pmap`.
+def jacobi_sym (a : ℤ) (b : ℕ) : ℤ :=
+(b.factors.pmap (λ p pp, @legendre_sym p ⟨pp⟩ a) (λ p pf, prime_of_mem_factors pf)).prod
+
+-- Notation for the Jacobi symbol.
+localized "notation `[` a ` | ` b `]ⱼ` := jacobi_sym a b" in number_theory_symbols
+
+/-!
+### Properties of the Jacobi symbol
+-/
+
+open_locale number_theory_symbols
+
+/-- The Jacobi symbol `(a / 0)` has the value `1`. -/
+lemma jacobi_sym_zero_right (a : ℤ) : [a | 0]ⱼ = 1 :=
+by simp only [jacobi_sym, factors_zero, list.prod_nil, list.pmap]
+
+/-- The Jacobi symbol `(a / 1)` has the value `1`. -/
+lemma jacobi_sym_one_right (a : ℤ) : [a | 1]ⱼ = 1 :=
+by simp only [jacobi_sym, factors_one, list.prod_nil, list.pmap]
+
+/-- The Legendre symbol `(a / p)` with an integer `a` and a prime number `p`
+is the same as the Jaocbi symbol `(a / p)`. -/
+lemma legendre_sym.to_jacobi_sym {p : ℕ} [fp : fact p.prime] {a : ℤ} :
+  legendre_sym p a = [a | p]ⱼ :=
+by simp only [jacobi_sym, factors_prime fp.1, list.prod_cons, list.prod_nil, mul_one, list.pmap]
+
+/-- The Jacobi symbol is multiplicative in its second argument. -/
+lemma jacobi_sym_mul_right (a : ℤ) (b₁ b₂ : ℕ) [ne_zero b₁] [ne_zero b₂] :
+  [a | b₁ * b₂]ⱼ = [a | b₁]ⱼ * [a | b₂]ⱼ :=
+begin
+  rw [jacobi_sym, ((perm_factors_mul (ne_zero.ne b₁) (ne_zero.ne b₂)).pmap _).prod_eq,
+      list.pmap_append, list.prod_append],
+  exacts [rfl, λ p hp, (list.mem_append.mp hp).elim prime_of_mem_factors prime_of_mem_factors],
+end
+
+/-- The Jacobi symbol takes only the values `0`, `1` and `1`. -/
+lemma jacobi_sym_trichotomy (a : ℤ) (b : ℕ) : [a | b]ⱼ = 0 ∨ [a | b]ⱼ = 1 ∨ [a | b]ⱼ = -1 :=
+begin
+  simp_rw [jacobi_sym],
+  let P : ℤ → Prop := λ x, x = 0 ∨ x = 1 ∨ x = -1,
+  have help : ∀ x₁ x₂ : ℤ, P x₁ → P x₂ → P (x₁ * x₂) :=
+  λ x₁ x₂ hx₁ hx₂,
+    by rcases hx₁ with (hx₁ | hx₁ | hx₁); rcases hx₂ with (hx₂ | hx₂ | hx₂);
+       simp_rw [P, hx₁, hx₂]; norm_num,
+  refine list.prod_prop (dec_trivial : P 1) (λ a' ha', _) help,
+  rcases list.mem_pmap.mp ha' with ⟨p, hp, hl⟩,
+  rw [← hl, legendre_sym],
+  haveI : fact p.prime := ⟨prime_of_mem_factors hp⟩,
+  exact quadratic_char_is_quadratic (zmod p) a,
+/-
+  sorry
+  {refine rec_on_mul _ _ (λ p pp, _) (λ b₁ b₂ h₁ h₂, _) b,
+  { rw jacobi_sym_zero_right, right, left, refl, },
+  { rw jacobi_sym_one_right, right, left, refl, },
+  { haveI : fact p.prime := ⟨pp⟩,
+    rw [← legendre_sym.to_jacobi_sym, legendre_sym],
+    exact quadratic_char_is_quadratic (zmod p) a, },
+  { by_cases hb₁: b₁ = 0, { rw [hb₁, zero_mul, jacobi_sym_zero_right], right, left, refl, },
+    by_cases hb₂: b₂ = 0, { rw [hb₂, mul_zero, jacobi_sym_zero_right], right, left, refl, },
+    rw [@jacobi_sym_mul_right _ _ _ ⟨hb₁⟩ ⟨hb₂⟩],
+    exact help h₁ h₂, } } -/
+end
```

Comment:
```suggestion
submonoid.list_prod_mem
  ⟨({0, 1, -1} : set ℤ),
   by rintro _ _ (rfl|rfl|(rfl : _ = _)) (rfl|rfl|(rfl : _ = _)); dec_trivial,
   or.inr (or.inl rfl)⟩
begin
  intros _ ha',
  rcases list.mem_pmap.mp ha' with ⟨p, hp, rfl⟩,
  haveI : fact p.prime := ⟨prime_of_mem_factors hp⟩,
  exact quadratic_char_is_quadratic (zmod p) a,
end
```
as I suggested on Zulip.

Reply:
OK (with the submonoid defined separately in the `int` section at the beginning).